File: services/sentra-detector-sidecar/src/sidecar/contracts.py
"""
Pydantic mirror of `@szl-holdings/sentra-detector-sdk`.

When the TS SDK changes, mirror those changes here. The api-server
validates incoming payloads with the zod schemas in
`packages/sentra-detector-sdk/src/schemas.ts`, so any drift between this
file and that one will surface immediately as a 400 on the sidecar
register / run round-trip.
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Awaitable, Callable, Literal, Optional, Protocol

from pydantic import BaseModel, Field

DetectorKind = Literal["heuristic", "signature", "statistical", "ml", "correlation"]
DetectorRuntime = Literal["ts", "python"]
CostClass = Literal["free", "cheap", "moderate", "expensive"]
GovernanceClass = Literal["read-only", "advisory", "mutating", "auto-remediable"]
FindingSeverity = Literal["critical", "high", "medium", "low", "info"]


class DetectorManifest(BaseModel):
    id: str
    label: str
    description: str
    kind: DetectorKind
    runtime: DetectorRuntime = "python"
    inputs: list[str] = Field(default_factory=list)
    costClass: CostClass
    governanceClass: GovernanceClass
    attackTechniques: Optional[list[str]] = None
    version: Optional[str] = None


class RecommendedAction(BaseModel):
    kind: Literal["patch", "block", "quarantine", "investigate", "tune"]
    detail: str


class Finding(BaseModel):
    id: str
    detectorId: str
    runId: str
    severity: FindingSeverity
    score: float = Field(ge=0.0, le=1.0)
    title: str
    summary: str
    attackTechniques: Optional[list[str]] = None
    affectedAssets: list[str] = Field(default_factory=list)
    evidence: dict[str, Any] = Field(default_factory=dict)
    recommendedAction: Optional[RecommendedAction] = None
    emittedAt: str
    governanceClass: GovernanceClass


class TraceLine(BaseModel):
    ts: str
    msg: str
    data: Optional[dict[str, Any]] = None


class SidecarRunRequest(BaseModel):
    detectorId: str
    runId: str
    triggeredBy: str = "system"
    startedAt: str
    params: dict[str, Any] = Field(default_factory=dict)
    inputs: dict[str, list[Any]] = Field(default_factory=dict)


class SidecarRunResponse(BaseModel):
    status: Literal["ok", "error"]
    findings: list[Finding] = Field(default_factory=list)
    trace: list[TraceLine] = Field(default_factory=list)
    errorMessage: Optional[str] = None


class SidecarRegisterRequest(BaseModel):
    sidecarId: str
    baseUrl: str
    detectors: list[DetectorManifest]


class DetectorContext(BaseModel):
    """Runtime context handed to a Python detector at evaluation time."""

    model_config = {"arbitrary_types_allowed": True}

    detectorId: str
    runId: str
    startedAt: str
    triggeredBy: str
    params: dict[str, Any]
    inputs: dict[str, list[Any]]
    trace_lines: list[TraceLine] = Field(default_factory=list)

    def read(self, input_name: str) -> list[Any]:
        return self.inputs.get(input_name, [])

    def trace(self, msg: str, data: Optional[dict[str, Any]] = None) -> None:
        self.trace_lines.append(
            TraceLine(
                ts=datetime.now(timezone.utc).isoformat(),
                msg=msg,
                data=data,
            )
        )


EvaluateFn = Callable[[DetectorContext], Awaitable[list[Finding]]]


class Detector(Protocol):
    manifest: DetectorManifest

    async def evaluate(self, ctx: DetectorContext) -> list[Finding]:  # pragma: no cover - protocol
        ...

File: packages/sentra-detector-sdk/src/contracts.ts
/**
 * Detector framework contracts.
 *
 * A `Detector` is anything that can be invoked with a `DetectorContext`
 * and emits zero or more `Finding`s. Detectors live in one of two homes:
 *
 *   - `runtime: 'ts'`     — TypeScript classes registered in-process
 *                           inside the api-server.
 *   - `runtime: 'python'` — Python classes hosted by the sidecar; the
 *                           api-server invokes them over HTTP via the
 *                           wire protocol in `./wire.ts` (mirrored in
 *                           `services/sentra-detector-sidecar`).
 *
 * Both runtimes return the SAME `Finding` shape so the existing alerts /
 * queue / workcell surfaces consume framework output without UI churn.
 */

/** Stable detector identifier (e.g. `ts-example/heuristic-port-scan`). */
export type DetectorId = string;

/** How sensitive the detector is to false positives / mis-calibration. */
export type CostClass = 'free' | 'cheap' | 'moderate' | 'expensive';

/**
 * Governance class — drives policy gating on the finding side:
 *
 *  - `read-only`        — the detector only inspects telemetry.
 *  - `advisory`         — the detector can request human review.
 *  - `mutating`         — the detector recommends action; a11oy gate
 *                         must approve before the workcell executes.
 *  - `auto-remediable`  — pre-approved action class, still receipted.
 */
export type GovernanceClass =
  | 'read-only'
  | 'advisory'
  | 'mutating'
  | 'auto-remediable';

/**
 * `antivenom` (#5503) — adversarial-input detectors that recognise the
 * counter-pattern of an attack instead of the attack itself (prompt
 * injection, jailbreak, model-extraction probing, adversarial-example
 * perturbations). Kept as a first-class kind, not a tag on `ml`, so the
 * Detector Council (MARBLE) can weight antivenom evidence distinctly
 * from baseline-shift evidence when arbitrating overlapping firings.
 */
export type DetectorKind =
  | 'heuristic'
  | 'signature'
  | 'statistical'
  | 'ml'
  | 'correlation'
  | 'antivenom'
  | 'temporal';

export type DetectorRuntime = 'ts' | 'python';

export type FindingSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info';

/**
 * Declarative description of a detector. Used by both the in-process TS
 * registry and the sidecar's registration handshake.
 */
export interface DetectorManifest {
  id: DetectorId;
  /** Short human label shown in the operator UI. */
  label: string;
  /** One-paragraph description shown in the run-history view. */
  description: string;
  kind: DetectorKind;
  runtime: DetectorRuntime;
  /** Free-form input descriptors (telemetry stream names, table names, …). */
  inputs: string[];
  costClass: CostClass;
  governanceClass: GovernanceClass;
  /** ATT&CK technique IDs the detector targets (e.g. ["T1046"]). */
  attackTechniques?: string[];
  /** Optional version tag — distinct deployments of the same id. */
  version?: string;
}

/**
 * Runtime context handed to a detector at evaluation time. Detectors must
 * pull their inputs through `read` instead of reaching into globals so
 * the framework can stub them in tests and account for IO in receipts.
 */
export interface DetectorContext {
  detectorId: DetectorId;
  runId: string;
  /** ISO-8601 — the wall-clock at run start. */
  startedAt: string;
  /** Operator / agent that triggered the run. */
  triggeredBy: string;
  /** Free-form run parameters (e.g. windowMinutes, baselineId). */
  params: Record<string, unknown>;
  /**
   * Read a logical input by name. The framework wires this through to
   * the appropriate telemetry source (db table, in-memory store, mock).
   */
  read: (input: string) => Promise<unknown[]>;
  /** Append a free-form trace line; surfaces in the run-history view. */
  trace: (msg: string, data?: Record<string, unknown>) => void;
}

/**
 * Canonical finding shape. Maps 1:1 onto the existing Sentra alert /
 * queue / workcell surfaces:
 *
 *   - alerts page consumes `severity`, `title`, `summary`, `evidence`.
 *   - queue/workcell consumes `recommendedAction`, `governanceClass`.
 *   - the receipt path consumes `chainReceiptId`.
 */
export interface Finding {
  /** Stable finding ID — `{detectorId}#{runId}#{idx}`. */
  id: string;
  detectorId: DetectorId;
  runId: string;
  severity: FindingSeverity;
  /** Numeric score 0..1 — higher is more suspicious. */
  score: number;
  title: string;
  summary: string;
  /** ATT&CK technique IDs this finding maps to. */
  attackTechniques?: string[];
  /** Asset references — flow into the existing affectedAssets fields. */
  affectedAssets: string[];
  /** Structured evidence dump (queryable rows, drift vectors, etc.). */
  evidence: Record<string, unknown>;
  /** Optional recommendation surfaced to the workcell. */
  recommendedAction?: {
    kind: 'patch' | 'block' | 'quarantine' | 'investigate' | 'tune';
    detail: string;
  };
  /** ISO-8601 emission time. */
  emittedAt: string;
  /** Governance class inherited from the detector that emitted it. */
  governanceClass: GovernanceClass;
}

/**
 * Per-run metadata recorded for the run-history view. Findings emitted
 * during the run reference the same `runId`.
 */
export interface DetectorRun {
  id: string;
  detectorId: DetectorId;
  startedAt: string;
  finishedAt: string;
  durationMs: number;
  status: 'ok' | 'error' | 'timeout';
  triggeredBy: string;
  findingsCount: number;
  /** SHA-256 chain receipt id from `@szl-holdings/szl-receipts`. */
  chainReceiptId?: string;
  errorMessage?: string;
  /** Free-form trace lines emitted via `ctx.trace`. */
  trace: Array<{ ts: string; msg: string; data?: Record<string, unknown> }>;
}

/**
 * The runtime contract a TypeScript detector implements. Python
 * detectors expose the equivalent shape over HTTP via `./wire.ts`.
 */
export interface Detector {
  readonly manifest: DetectorManifest;
  evaluate(ctx: DetectorContext): Promise<Finding[]>;
}

/**
 * Severity threshold at which the framework fires
 * `crossProductHandoff` to A11oy. Centralised here so any consumer can
 * agree on what counts as "high-enough to escalate".
 */
export const A11OY_HANDOFF_THRESHOLD: FindingSeverity = 'high';
export const HANDOFF_RANK: Record<FindingSeverity, number> = {
  info: 0,
  low: 1,
  medium: 2,
  high: 3,
  critical: 4,
};

export function shouldHandoff(severity: FindingSeverity): boolean {
  return HANDOFF_RANK[severity] >= HANDOFF_RANK[A11OY_HANDOFF_THRESHOLD];
}
